staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go

```diff
@@ -295,112 +395,83 @@ func (cnc *CloudNodeController) updateNodeAddress(ctx context.Context, node *v1.
-				LastTransitionTime: metav1.Now(),
-			}); err != nil {
-				return err
-			}
-		}
-		return nil
-	})
+// syncNode handles updating existing nodes registered with the cloud taint
+// and processes nodes that were added into the cluster, and cloud initialize them if appropriate.
+func (cnc *CloudNodeController) syncNode(ctx context.Context, nodeName string) error {
+	curNode, err := cnc.nodeInformer.Lister().Get(nodeName)
 	if err != nil {
-		utilruntime.HandleError(err)
-		return
-	}
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
 
-	curNode, err := cnc.kubeClient.CoreV1().Nodes().Get(context.TODO(), node.Name, metav1.GetOptions{})
-	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("failed to get node %s: %v", node.Name, err))
-		return
+		return err
 	}
 
 	cloudTaint := getCloudTaint(curNode.Spec.Taints)
 	if cloudTaint == nil {
 		// Node object received from event had the cloud taint but was outdated,
-		// the node has actually already been initialized.
-		return
+		// the node has actually already been initialized, so this sync event can be ignored.
+		return nil
 	}
 
+	klog.Infof("Initializing node %s with cloud provider", nodeName)
+
 	// TODO: getNodeModifiersFromCloudProvider and updateNodeAddress both call cloud api to get instanceMetadata,
 	// get instanceMetadata and pass it to getNodeModifiersFromCloudProvider and updateNodeAddress which reduces api calls.
-	nodeModifiers, err := cnc.getNodeModifiersFromCloudProvider(ctx, curNode)
+	copyNode := curNode.DeepCopy()
+	nodeModifiers, err := cnc.getNodeModifiersFromCloudProvider(ctx, copyNode)
 	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("failed to initialize node %s at cloudprovider: %v", node.Name, err))
-		return
+		return fmt.Errorf("failed to get node modifiers from cloud provider: %v", err)
 	}
 
 	nodeModifiers = append(nodeModifiers, func(n *v1.Node) {
 		n.Spec.Taints = excludeCloudTaint(n.Spec.Taints)
 	})
 
 	err = clientretry.RetryOnConflict(UpdateNodeSpecBackoff, func() error {
-		curNode, err := cnc.kubeClient.CoreV1().Nodes().Get(context.TODO(), node.Name, metav1.GetOptions{})
+		curNode, err := cnc.nodeInformer.Lister().Get(nodeName)
 		if err != nil {
 			return err
 		}
 
+		newNode := curNode.DeepCopy()
 		for _, modify := range nodeModifiers {
-			modify(curNode)
+			modify(newNode)
 		}
 
-		_, err = cnc.kubeClient.CoreV1().Nodes().Update(context.TODO(), curNode, metav1.UpdateOptions{})
+		_, err = cnc.kubeClient.CoreV1().Nodes().Update(context.TODO(), newNode, metav1.UpdateOptions{})
 		if err != nil {
 			return err
 		}
 
+		if cnc.cloud.ProviderName() == "gce" {
+			// TODO(wlan0): Move this logic to the route controller using the node taint instead of condition
+			// Since there are node taints, do we still need this?
+			// This condition marks the node as unusable until routes are initialized in the cloud provider
+			if err := cloudnodeutil.SetNodeCondition(cnc.kubeClient, types.NodeName(nodeName), v1.NodeCondition{
```

Comment:
I think we actually want the GCE network unavailable condition to happen before the Update on line 442, since we want the condition in place before the cloud provider init taint is removed. Otherwise there will be a brief period where the node doesn't have the cloud taint and the condition.

Reply:
@andrewsykim The unit tests would fail if I move GCE network unavailable condition to happen before the **Update**
Not sure why it causes the error, still investigating 🤔